plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -0,0 +1,645 @@
+{
+	const gchar *attribute;
+	guint64 status;
+
+	/* now check if the update actually worked */
+	attribute = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (device),
+						   "nvm_authenticate",
+						   error);
+	if (attribute == NULL)
+		return FALSE;
+	status = g_ascii_strtoull (attribute, NULL, 16);
+	if (status == G_MAXUINT64 && errno == ERANGE) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "failed to read 'nvm_authenticate: %s",
+			     g_strerror (errno));
+		return FALSE;
+	}
+
+	/* anything else then 0x0 means we got an error */
+	if (status != 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "update failed (status %" G_GINT64_MODIFIER "x)", status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_write_data (FuThunderboltDevice	*self,
+				  GBytes		*blob_fw,
+				  GError		**error)
+{
+	gsize fw_size;
+	gsize nwritten;
+	gssize n;
+	g_autoptr(GFile) nvmem = NULL;
+	g_autoptr(GOutputStream) os = NULL;
+
+	nvmem = fu_thunderbolt_device_find_nvmem (self, FALSE, error);
+	if (nvmem == NULL)
+		return FALSE;
+
+	os = (GOutputStream *) g_file_append_to (nvmem,
+						 G_FILE_CREATE_NONE,
+						 NULL,
+						 error);
+
+	if (os == NULL)
+		return FALSE;
+
+	nwritten = 0;
+	fw_size = g_bytes_get_size (blob_fw);
+	fu_device_set_progress_full (FU_DEVICE (self), nwritten, fw_size);
+
+	do {
+		g_autoptr(GBytes) fw_data = NULL;
+
+		fw_data = g_bytes_new_from_bytes (blob_fw,
+						  nwritten,
+						  fw_size - nwritten);
+
+		n = g_output_stream_write_bytes (os,
+						 fw_data,
+						 NULL,
+						 error);
+		if (n < 0)
+			return FALSE;
+
+		nwritten += n;
+		fu_device_set_progress_full (FU_DEVICE (self), nwritten, fw_size);
+
+	} while (nwritten < fw_size);
+
+	if (nwritten != fw_size) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "Could not write all data to nvmem");
+		return FALSE;
+	}
+
+	return g_output_stream_close (os, NULL, error);
+}
+
+static FuFirmware *
+fu_thunderbolt_device_prepare_firmware (FuDevice *device,
+					GBytes *fw,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE (device);
+	g_autoptr(FuThunderboltFirmware) firmware = fu_thunderbolt_firmware_new ();
+	g_autoptr(FuThunderboltFirmware) firmware_old = fu_thunderbolt_firmware_new ();
+	g_autoptr(GBytes) controller_fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GFile) nvmem = NULL;
+
+	/* parse */
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware), fw, flags, error))
+		return NULL;
+
+	/* get current NVMEM */
+	nvmem = fu_thunderbolt_device_find_nvmem (self, TRUE, error);
+	if (nvmem == NULL)
+		return NULL;
+	controller_fw = g_file_load_bytes (nvmem, NULL, NULL, error);
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware_old), controller_fw, flags, error))
+		return NULL;
+	if (fu_thunderbolt_firmware_is_host (firmware) !=
```

Comment:
Since there is a way to tell if it's host by reading the FW image itself now - maybe we should use that at startup rather than the examining of the udev parent?  This seems safer.

Reply:
Is we want to find out if it's host we have to read a *lot* more of the image. 

    failed to read is-host: location is outside of the given image: attempted to read 0x01 bytes at offset 0x7f014b from buffer of 0x80

I've pushed a commit that at least makes use of the native boolean.

Reply:
Sounds good, thanks.